### ISSUE TITLE
Board is transposed on Black's turn, not mirrored like it should be

### ISSUE THREAD

**Issue by Levi-Lesches** [Learning-Python-Team/Chess#11]
Basically, your pieces are always shown at the bottom. So moving, say, a2 --> a3 should be valid when your pieces are actually at a2, but not when they're simply printed that way.

So what should we do? Keep the display of the board the same, no matter who's turn it is? Or have a function to translate a2 to g7 if needed?

**Comment by ZeroOne010101** [comment]
ok, i havent yet looked at the new draw function, but i assume the numbers rotate with it like in the old one?
in that case the moves are still valid.
so whites bottom left is a1 and blacks bottom left is a8 so when black moves its rook, it moves say a8 a7 which is a valid move, same for white. 

The trick comes when you need to check if the move is valid, since only the vectors from whites perspective are in the pieces class. so what i do is in my move_valid prototype i reverse the vectors for black like this:
```
        if players.turn == players.white:
            vector = (destination_row - origin_row, destination_column - origin_column)
        else:
            vector = (destination_row - origin_row, destination_column - origin_column)
            vector = (-1 * vector[0], -1 * vector[1])
```
making them match the white vectors. since move_valid just tests and not actually moves this works :)

**Comment by ZeroOne010101** [comment]
here, i printed the old version to visualise it:
python chess.py 
```
  |
8 |  ♖    ♘    ♗    ♕    ♔    ♗    ♘    ♖  
  |
7 |  ♙    ♙    ♙    ♙    ♙    ♙    ♙    ♙  
  |
6 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
5 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
4 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
3 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
2 |  ♟    ♟    ♟    ♟    ♟    ♟    ♟    ♟  
  |
1 |  ♜    ♞    ♝    ♛    ♚    ♝    ♞    ♜  
¯¯T¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
  |  a    b    c    d    e    f    g    h
white: a2 a3
  |
1 |  ♜    ♞    ♝    ♛    ♚    ♝    ♞    ♜  
  |
2 |  ☐    ♟    ♟    ♟    ♟    ♟    ♟    ♟  
  |
3 |  ♟    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
4 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
5 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
6 |  ☐    ☐    ☐    ☐    ☐    ☐    ☐    ☐  
  |
7 |  ♙    ♙    ♙    ♙    ♙    ♙    ♙    ♙  
  |
8 |  ♖    ♘    ♗    ♕    ♔    ♗    ♘    ♖  
¯¯T¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯¯
  |  a    b    c    d    e    f    g    h
black: 
```

**Comment by Levi-Lesches** [comment]
Yes, the draw function keeps the `reversed(self.game_board)` part. But the numbers are not re-oredered. 

Also, have you changed your move logic? Because this does not happen on my copy

**Comment by ZeroOne010101** [comment]
not that i knew... ill have a quick look at your fork

**Comment by Levi-Lesches** [comment]
Oh OK I see now. I was confused by the fact that the board isn't mirrored, it's transposed

**Comment by ZeroOne010101** [comment]
yes, its supposed to be that way. just imagine a chessboard in real life, the black player is always at the side of the 8, and the white player at the side with the . its a change of perspective from the white players to the black players

**Comment by Levi-Lesches** [comment]
Right but a and f switch around

**Comment by ZeroOne010101** [comment]
not in the chessboards that i know of. 
https://en.wikipedia.org/wiki/Chess look at the board they have in the setup section . a is a and h is h

**Comment by Levi-Lesches** [comment]
Ok, so TIL the letters don't change, but say for black, the king is on the right, then for white, the king is on the left, no?

**Comment by ZeroOne010101** [comment]
correct, whites left is blacks right, since the players are sitting opposite to each other.

**Comment by Levi-Lesches** [comment]
right, but that's not printed

**Comment by ZeroOne010101** [comment]
shit, ur right. fixing the draw function should suffice since e on the left is still e.

**Comment by ZeroOne010101** [comment]
so well need to pass players to board.draw so it can read players.turn

**Comment by ZeroOne010101** [comment]
in other words, we do need to mirror it, but with the correct nr's and characters.

**Comment by Levi-Lesches** [comment]
Yeah, I'll get to work on it. The good news is that `reversed` should do the majority of the work here, assuming a list that goes in order, so `[a1, b1,...h1, a1, b2...h2...]` should mirror properly. If we just construct that list and pass it into `reversed` and then deconstruct it into a list of lists again, it should work.

**Comment by webdotorg** [comment]
Sorry I missed that before.

**Comment by Levi-Lesches** [comment]
Ok, so I got it worked out, I'll submit a pull request as soon as it works again. But there's a problem in `Board.move`. Typing in `a2 a3` translates to `( (1, 0), (2, 0) )`, but that really means moving from top to bottom, since 0 refers to the top row (from white's perspective). 

I'll open a separate issue for that and take care of it.

**Comment by webdotorg** [comment]
Sorry I missed that before. I just noticed something else that will likely come into play while fixing this bug.
Standard chess rules dictate that white always goes first, placing the black square in the bottom left-hand corner (a1). After white moves, black is now in white's corner.
This can be confusing or at least problematic for the proposed future AI function.

**Comment by Levi-Lesches** [comment]
Sorry, what do you mean, "the black square"?

**Comment by Levi-Lesches** [comment]
#12 is about this new issue, by the way.

**Comment by Levi-Lesches** [comment]
Ready to merge?

**Comment by Levi-Lesches** [comment]
Solved with #13
